File: mpi-1x-hypack-2013/matrix-vector-matrix-comp-codes/f77-lang-matvect/mm_mult_master_self_schd.f
c
c*******************************************************************
c
c		C-DAC Tech Workshop : hyPACK-2013
c                           October 15-18, 2013
c
c  Example 5.8		: mm_mult_master_sschd.f
c
c  Objective           : Matrix_Matrix_Multiplication
c                        (Self_Scheduling algorithm master program)
c
c  Input               : Process 0 (master) reads files (mdata1.inp) 
c                        for Matrix_A and (mdata2.inp) for Matrix_B.
c
c  Output              : Process 0 prints the result of Matrix-Matrix 
c                        Multiplication 
c
c  Necessary Condition : Number of Processes should be less than 
c                        or equal to 8
c
c   Created           : August-2013
c
c   E-mail            : hpcfte@cdac.in     
c
c*******************************************************************

      program main

      implicit none

      include 'mpif.h'

      integer MAX_ROWS, MAX_COLS
      parameter (MAX_ROWS = 100, MAX_COLS = 100)
c

      integer Dest, RowtoSend
      integer NoofRowsA, NoofColsA
      integer NoofRowsB, NoofColsB

      double precision Matrix_A(MAX_ROWS,MAX_COLS), 
     $	               Matrix_B(MAX_ROWS,MAX_COLS), 
     $                 ResultMatrix(MAX_ROWS,MAX_COLS)
      double precision RowA(MAX_COLS),  ResultRow(MAX_COLS)

      real*8 Starttime, Endtime, Time
c
      integer MyRank, Root, Numprocs
      integer status(MPI_STATUS_SIZE)
      integer i, j, k, iproc, ierror
      integer rowtype, anstype, donetype
      integer Source_tag, Destination_tag
c
c      open(unit = 23, file = 'mat.out')

      call MPI_INIT(ierror)
      call MPI_COMM_RANK(MPI_COMM_WORLD, MyRank, ierror )
      call MPI_COMM_SIZE(MPI_COMM_WORLD, Numprocs, ierror )
c
      rowtype  = 1
      anstype  = 2
      donetype = 0
      Root     = 0

c     Initialize Matrix_A and Matrix_B

      Starttime = MPI_WTIME()

C        .......Read the Matrix From Input file ......
	 open(unit=12, file = './data/mdata1.inp')
         read(12,*) NoofRowsA, NoofColsA
         do i = 1,NoofRowsA
            read(12,*) (Matrix_A(i,j),j=1,NoofColsA)
C            write(6,*) (Matrix_A(i,j),j=1,NoofColsA)
         enddo

	 open(unit=13, file = './data/mdata2.inp')
         read(13,*) NoofRowsB, NoofColsB
         do i = 1,NoofRowsB
            read(13,*) (Matrix_B(i,j),j=1,NoofColsB)
C            write(6,*) (Matrix_B(i,j),j=1,NoofColsB)
         enddo

 	 close(12)
	 close(13)


	call MPI_Barrier(MPI_COMM_WORLD, ierror)
	call MPI_Bcast(NoofColsA, 1, MPI_INTEGER, Root,
     $ 	               MPI_COMM_WORLD,ierror)

	call MPI_Bcast(NoofRowsB, 1, MPI_INTEGER, Root,
     $  	       MPI_COMM_WORLD,ierror)

	if(NoofColsA .ne. NoofRowsB) then
  	 print*,"Incompatible dimensions of Matrices for Mutlipication"
   	 goto 100
        endif

	call MPI_Bcast(NoofColsB, 1, MPI_INTEGER, Root, 
     $ 	               MPI_COMM_WORLD,ierror)

c        send Matrix_B to each other process
	 do 86 i = 1, NoofColsB
          call MPI_BCAST(Matrix_B(1,i),NoofRowsB,MPI_DOUBLE_PRECISION, 
     $          Root,MPI_COMM_WORLD,ierror)
86	continue
c
c        send a row of Matrix_A to each other process
c
	 RowtoSend = 0
         do 40 iproc = 1, Numprocs-1
            do 30 j = 1,NoofColsA
               RowA(j) = Matrix_A(iproc,j)
 30         continue
c
            Dest = iproc
               call MPI_SEND(RowA, NoofColsA, MPI_DOUBLE_PRECISION,
     $              Dest, RowtoSend+1, MPI_COMM_WORLD, ierror)
               RowtoSend = RowtoSend+1
 40      continue
c
         do 70 i = 1, NoofRowsA
          call MPI_RECV( ResultRow, NoofColsA, MPI_DOUBLE_PRECISION,
     $      MPI_ANY_SOURCE,MPI_ANY_TAG,MPI_COMM_WORLD,status,ierror)
c
             Dest = status(MPI_SOURCE)
             Source_tag = status(MPI_TAG)
c
             do k = 1, NoofColsA 
	             ResultMatrix(Source_tag,k) =  ResultRow(k)
             end do
c
               if(RowtoSend .lt. NoofRowsA) then
               do 50 j = 1,NoofColsA
                  RowA(j) = Matrix_A(RowtoSend+1,j)
 50            continue
c
	     Destination_tag = RowtoSend + 1
             call MPI_SEND(RowA,NoofColsA,MPI_DOUBLE_PRECISION,Dest,
     $                       Destination_tag, MPI_COMM_WORLD,ierror)
               
               RowtoSend     = RowtoSend+1
            else
               RowA(1) = 1.0d0
               call MPI_SEND(RowA, 1, MPI_DOUBLE_PRECISION, 
     $                     Dest, donetype, MPI_COMM_WORLD, ierror)
            endif
	Endtime = MPI_Wtime()
	Time    = EndTime - Starttime
 70      continue
         
c      print out the answer
       write(6,75)((ResultMatrix(i,j), j=1, NoofColsA), i=1, NoofRowsA)
	print*, "  Time = ", Time

 100  	call MPI_FINALIZE(ierror)
75	   format(8(2x,f8.3))
      stop
      end

File: mpi-1x-hypack-2013/matrix-vector-matrix-comp-codes/c-lang-matvect/mm_mult_master_self_schd.c
/*
*******************************************************************

		C-DAC Tech Workshop : hyPACK-2013
                           October 15-18, 2013

  Example 5.8	      : mm_mult_master_self_schd.c

  Objective           : Matrix_Matrix_Multiplication
                        (Self_Scheduling-Master Program)

  Input               : Read files (mdata1.inp) for Matrix 1
                        & (mdata2.inp) for Matrix 2.

  Output              : Result of Matrix-Matrix Multiplication on process 0.

  Necessary Condition : Number of Processes should be
                        less than or equal to 8

   Created            : August-2013

   E-mail             : hpcfte@cdac.in     

*****************************************************************/

#include <stdio.h>
#include <stdlib.h>
#include <mpi.h>
#include <string.h>

main(int argc, char** argv) 
{

 int	     Numprocs, MyRank;
 int		  NoofRowsA, NoofColsA,	NoofRowsB, NoofColsB;
 int		  Source, Destination, Root=0;
 int	     Source_tag, Destination_tag;
 int 	     irow, icol, iproc, index;
 int	     RowtoSend = 0, RowReceived = 0;
 float	     **Matrix_A, **Matrix_B, **ResultMatrix;
 float	     *BufferB, *RowA, *ResultRow;
 FILE	     *fp;
 int		  MatrixA_FileStatus = 1, MatrixB_FileStatus = 1;
 MPI_Status status;     

 /* ........MPI Initialisation .......*/
	MPI_Init(&argc, &argv); 
  	MPI_Comm_rank(MPI_COMM_WORLD, &MyRank);
  	MPI_Comm_size(MPI_COMM_WORLD, &Numprocs);

 /* Reading Input */

    if ((fp = fopen ("./data/mdata1.inp", "r")) == NULL){
			MatrixA_FileStatus = 0;
     }

  if(MatrixA_FileStatus != 0) {
     fscanf(fp, "%d %d\n", &NoofRowsA, &NoofColsA);

     Matrix_A = (float **) malloc (NoofRowsA * sizeof(float *));
     for (irow = 0; irow < NoofRowsA; irow++){
	Matrix_A[irow] = (float *) malloc(NoofColsA * sizeof(float));
     		for (icol = 0; icol < NoofColsA; icol++)
       		fscanf(fp, "%f", &Matrix_A[irow][icol]);
     		}
     		fclose(fp);
	 }

     if((fp = fopen ("./data/mdata2.inp", "r")) == NULL){
			MatrixB_FileStatus = 0;
     }

  if(MatrixB_FileStatus != 0) {
	fscanf(fp, "%d %d\n", &NoofRowsB, &NoofColsB);

     	Matrix_B = (float **) malloc (NoofRowsB * sizeof(float *));
     for(irow = 0; irow < NoofRowsB; irow++){
      Matrix_B[irow] = (float *) malloc(NoofColsB * sizeof(float *));
       		for(icol = 0; icol < NoofColsB; icol++)
     		fscanf(fp, "%f", &Matrix_B[irow][icol]);
    		}
     		fclose(fp);
	  }

   MPI_Barrier(MPI_COMM_WORLD);
   MPI_Bcast (&MatrixA_FileStatus, 1, MPI_INT, 0, MPI_COMM_WORLD);
   if(MatrixA_FileStatus == 0) {
       printf("Can't open input file for Matrix A ......");
		 MPI_Finalize();
       exit(-1);
   }
   MPI_Bcast (&MatrixB_FileStatus, 1, MPI_INT, 0, MPI_COMM_WORLD);
   if(MatrixB_FileStatus == 0) {
       printf("Can't open input file for Matrix B ......");
		 MPI_Finalize();
       exit(-1);
   }

	MPI_Bcast (&NoofColsA, 1, MPI_INT, 0, MPI_COMM_WORLD);
	MPI_Bcast (&NoofRowsB, 1, MPI_INT, 0, MPI_COMM_WORLD);

   if(NoofColsA != NoofRowsB){
    MPI_Finalize();
    printf("Incompatible dimensions of Matrices for Mutlipication\n");
    exit(0);
 }

   MPI_Bcast (&NoofColsB, 1, MPI_INT, 0, MPI_COMM_WORLD);

 /* .......Convert Matrix B into 1-D BufferB and Broadcast to Workers.....*/
 BufferB = (float *) malloc(NoofRowsB*NoofColsB*sizeof(float));
 index  = 0;
 for(irow = 0; irow<NoofRowsB; irow++)
   for(icol=0; icol<NoofColsB; icol++)
   BufferB[index++] = Matrix_B[irow][icol];
   MPI_Bcast (BufferB,NoofRowsB*NoofColsB,MPI_FLOAT,0,MPI_COMM_WORLD);

 /* .......Allocate memory for output and a row of a Matrix A......*/ 
   RowA = (float *)malloc(NoofColsA*sizeof(float));
   ResultMatrix = (float **)malloc(NoofRowsA*sizeof(float *)); 
   for(irow=0; irow<NoofRowsA; irow++)
     ResultMatrix[irow] = (float *) malloc(NoofColsB*sizeof(float));

   /* First Send At least one row to each processors */
    RowtoSend = 0;
    for(iproc = 1 ; iproc < Numprocs ; iproc++) {
	for(icol=0; icol<NoofColsA; icol++)
	RowA[icol] = Matrix_A[RowtoSend][icol];
   	MPI_Send(RowA, NoofColsA, MPI_FLOAT, iproc, RowtoSend+1, MPI_COMM_WORLD);
	RowtoSend++;
	}

   ResultRow = (float *) malloc(NoofColsB*sizeof(float));
	for(irow = 0 ; irow < NoofRowsA ; irow++) {
	MPI_Recv(ResultRow, NoofColsB, MPI_FLOAT, MPI_ANY_SOURCE, MPI_ANY_TAG, MPI_COMM_WORLD, &status);
	Destination = status.MPI_SOURCE;
	Source_tag  = status.MPI_TAG;
		 
	memcpy(ResultMatrix[Source_tag], ResultRow, NoofColsB*sizeof(float));

 /*For the  remaining rows(if any) send the data to whichever processor that finishes the earlier computations first */

   if( RowtoSend < NoofRowsA ) {
	 memcpy(RowA, Matrix_A[RowtoSend], NoofColsA*sizeof(float));
 	 Destination_tag = RowtoSend+1;
   	 MPI_Send(RowA, NoofColsA, MPI_FLOAT, Destination, Destination_tag, MPI_COMM_WORLD);
	 RowtoSend++;
    }
	}

 for(iproc = 1 ; iproc < Numprocs ; iproc++) 
   MPI_Send(RowA, NoofColsA, MPI_FLOAT, iproc, 0, MPI_COMM_WORLD);
	
  printf("Result is ...\n");
	for(irow = 0 ; irow < NoofRowsA; irow++) {
		for( icol= 0 ;  icol< NoofColsB; icol++) 
			printf("  %3.3f   ",ResultMatrix[irow][icol]); 
	   printf("\n");
	}
		
	/*...Free Allocated Memory......*/
   for(irow=0; irow<NoofRowsA; irow++)
		free(Matrix_A[irow]);
   for(irow=0; irow<NoofRowsB; irow++)
		free(Matrix_B[irow]);
   for(irow=0; irow<NoofRowsA; irow++)
		free(ResultMatrix[irow]);

   free(Matrix_A);
   free(Matrix_B);
   free(ResultMatrix);
   free(RowA);
   free(BufferB);
   free(ResultRow);
   fclose(fp);

   MPI_Finalize(); 
}




